File: day_03/2.rb
INPUT = "./day_02/input"
SMALL_INPUT = "./day_02/small_input"

def get_most_common(filtered, pos)
  ones = filtered.reduce(0) { |sum, line| sum + Integer(line[pos]) }
  return ones >= len(filtered) / 2 ? "1" : "0"
end

lines = File.readlines(INPUT)

length = lines[0].length
filtered = lines
pos = 0

while filtered.length > 1 && pos < length
  common = get_most_common(filtered, pos)
  filtered = filtered.filter { |line| line[pos] == common }
  pos += 1
end

oxygen = filtered[0]

filtered = lines
pos = 0

while filtered.length > 1 && pos < length
  common = get_most_common(filtered, pos)
  common = common == "0" ? "1" : "0"

  filtered = filtered.filter { |line| line[pos] == common }
  pos += 1
end

co2 = filtered[0]

puts "Oxygen: #{oxygen}"
puts "CO2: #{co2}"

File: day_03/2.py
from typing import List
from functools import reduce


SMALL_INPUT: str = "./day_03/small_input";
INPUT: str = "./day_03/input";

def get_most_common(filtered: List[str], pos: int) -> str:
  ones: int = reduce( lambda acc, item: acc + int(item[pos]), filtered, 0)
  return "1" if ones >= len(filtered) / 2 else "0"

with open(INPUT) as f:
  lines: List[str] = f.readlines()

  length: int = len(lines[0])
  filtered: List[str] = lines
  pos: int = 0

  while len(filtered) > 1 and pos < length:
    common: str = get_most_common(filtered, pos)
    filtered = [l for l in filtered if l[pos] == common]
    pos += 1

  oxygen = filtered[0].strip("\n")

  filtered: List[str] = lines
  pos: int = 0

  while len(filtered) > 1 and pos < length:
    common: str = get_most_common(filtered, pos)
    common = "1" if common == "0" else "0"

    filtered = [l for l in filtered if l[pos] == common]
    pos += 1

  co2 = filtered[0].strip("\n")

  print(f"Oxygen: {oxygen}")
  print(f"CO2: {co2}")
  print(f"things: {int(oxygen, 2) * int(co2, 2)}")

